File: vision/classify.py
"""Risk classification and series aggregation. threshold_version = th-1.0.0

Prototype heuristics. Not Tamir 2017. Not a clinical cutoff.
"""

from __future__ import annotations

from statistics import median
from typing import Optional, Sequence

HIGH = "HIGH"
MODERATE = "MODERATE"
LOW = "LOW"
UNABLE = "UNABLE_TO_ASSESS"


def classify_anemia(a_star: float) -> str:
    if a_star < 5.0:
        return HIGH
    if a_star < 10.0:
        return MODERATE
    return LOW


def classify_jaundice(b_star: float) -> str:
    if b_star >= 15.0:
        return HIGH
    if b_star >= 10.0:
        return MODERATE
    return LOW


def aggregate(values: Sequence[float]) -> tuple[str, Optional[float]]:
    """< 2 valid → UNABLE, signal None. n=2 median is the arithmetic mean."""
    valid = [float(v) for v in values]
    if len(valid) < 2:
        return UNABLE, None
    return "", float(median(valid))


def classify_anemia_series(values: Sequence[float]) -> tuple[str, Optional[float]]:
    _, signal = aggregate(values)
    if signal is None:
        return UNABLE, None
    return classify_anemia(signal), signal


def classify_jaundice_series(values: Sequence[float]) -> tuple[str, Optional[float]]:
    _, signal = aggregate(values)
    if signal is None:
        return UNABLE, None
    return classify_jaundice(signal), signal

File: app/lib/src/vision/classify.dart
/// Prototype heuristics. threshold_version = th-1.0.0. Not Tamir 2017.
enum Risk { low, moderate, high, unableToAssess }

String riskWire(Risk r) => switch (r) {
      Risk.low => 'LOW',
      Risk.moderate => 'MODERATE',
      Risk.high => 'HIGH',
      Risk.unableToAssess => 'UNABLE_TO_ASSESS',
    };

Risk classifyAnemia(double aStar) {
  if (aStar < 5.0) return Risk.high;
  if (aStar < 10.0) return Risk.moderate;
  return Risk.low;
}

Risk classifyJaundice(double bStar) {
  if (bStar >= 15.0) return Risk.high;
  if (bStar >= 10.0) return Risk.moderate;
  return Risk.low;
}

double? seriesMedian(List<double> values) {
  if (values.length < 2) return null;
  final s = [...values]..sort();
  final n = s.length;
  if (n.isOdd) return s[n ~/ 2];
  return (s[n ~/ 2 - 1] + s[n ~/ 2]) / 2.0;
}

({Risk risk, double? signal}) classifyAnemiaSeries(List<double> values) {
  final signal = seriesMedian(values);
  if (signal == null) return (risk: Risk.unableToAssess, signal: null);
  return (risk: classifyAnemia(signal), signal: signal);
}

({Risk risk, double? signal}) classifyJaundiceSeries(List<double> values) {
  final signal = seriesMedian(values);
  if (signal == null) return (risk: Risk.unableToAssess, signal: null);
  return (risk: classifyJaundice(signal), signal: signal);
}
